src/vs/base/common/strings.ts

```diff
@@ -1361,3 +1361,30 @@ export class InvisibleCharacters {
 }
 
 export const Ellipsis = '\u2026';
+
+/**
+ * Convert a Unicode string to a string in which each 16-bit unit occupies only one byte
+ *
+ * From https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/btoa
+ */
+function toBinary(str: string): string {
+	const codeUnits = new Uint16Array(str.length);
+	for (let i = 0; i < codeUnits.length; i++) {
+		codeUnits[i] = str.charCodeAt(i);
+	}
+	let binary = '';
+	const uint8array = new Uint8Array(codeUnits.buffer);
+	for (let i = 0; i < uint8array.length; i++) {
+		binary += String.fromCharCode(uint8array[i]);
+	}
+	return binary;
+}
+
+/**
+ * Version of the global `btoa` function that handles multi-byte characters instead
+ * of throwing an exception.
+ */
+
+export function multibyteAwareBtoa(str: string): string {
```

Comment:
We have `encodeBase64` already, do we need a separate btoa function?

Reply:
A bit nuanced. The two functions produce different outputs and have subtly different behavior around the output being url safe or not by default

I think we could replace calls to `multibyteAwareBtoa` with `encodeBase64(VSBuffer.fromString(str), true, true)` but that's too verbose. I think we should keep this version but will see if the implementation can just make it call `encodeBase64`